Adding document for comment (#18)
bytecodealliance/sig-embedded: ExploringOptions/Alternative_Solutions_(WASI_Wasm).md

> # WASI / Wasm System Structure Exploration of improvements or alternative solutions

Reviewer: @woodsmc, thanks for moving this to a separate PR &mdash; it's a lot easier to review. Have you already dealt with the issues brought up in https://github.com/bytecodealliance/sig-embedded/issues/10 or were you planning to get to that at some point? (I just didn't want that valuable feedback lost in the move!).

Author: Thanks for reminding me!- That is very much appreciated. For completeness I've copied @lukewagner's original issue text from [here](https://github.com/bytecodealliance/sig-embedded/issues/10) and I've inlined my responses from that thread. The two edits suggested in my responses have been applied ( I know these got a little lost in the transfer to the PR, I appreciate everyone's patience with this). To help the reader, my original responses from the original issue are prefaced with "As noted:". 

> 
> Unfortunately #9 was not kept open to allow feedback; perhaps we could do that in the future since it makes it easier to have threaded discussions.  

**As noted:** Good call. Wasn't intentional and good feedback.

> In any case, there are some pretty significant inaccuracies in the Introduction, Problem Definition and Status Quo sections, which I'll quote and describe:
> 
> > An initial release candidate for WASI 1.0 is expected to be released by the end of the year. However, based on current implementations and the publicly shared roadmap, WASI 1.0 will be largely unsuitable for embedded applications.
> 
> I don't know where this comes from, but it's quite untrue.  The timeline presented at the [last two CG meetings](https://docs.google.com/presentation/d/1z0WXS5BLFtbVynM9xAyilecYskN1IKe9Dad1nDEmgU8/edit#slide=id.g33067d21cc1_0_321) indicates a 1.0 "release candidate" no earlier than mid-2026.  And then further work is required to (1) split off a separate W3C Working Group for WASI, (2) advance stages from "release candidate" to actual standard 1.0.  

**As noted:** Yup, let's update the document to reflect that. This now reads:
```
As of early 2025, the WASI 0.2 release is available, with 0.3 expected in the coming months. An initial release candidate for WASI 1.0 is expected to be released by no later than mid-2026, at which point there is a desire to seperate the standardization process by the creation of a new W3C charter - [link with details]([Wasm CG Feburary 2025 - Google Slides](https://docs.google.com/presentation/d/1z0WXS5BLFtbVynM9xAyilecYskN1IKe9Dad1nDEmgU8/edit#slide=id.g33067d21cc1_0_321)).
```

> Thus, it's far too early to speculate on whether or not WASI 1.0 will be suitable for embedded applications and so I think this paragraph should simply be removed.

> > A significant challenge is the lack of a viable migration path for existing WASI Preview 1 runtimes. [...]
> 
> There are several migration paths being discussed at this moment with folks in this SIG.  One of the most promising, whose **viability** is already demonstrated by [jco](https://github.com/bytecodealliance/jco), involves preprocessing component binaries into core modules that can be run with modest additional effort by existing core wasm runtimes.  Extending this approach to more core runtimes like WAMR seems quite viable, and so the quoted sentence and much of the following paragraph seems incorrect.  I think it would be more appropriate to explain the current challenge and say that solutions exist that would require far less engineering effort than implementing the Component Model from scratch.

**As noted:** To be 100% transparent with everyone we should state the difference between feasible, viable and concrete reality. Jco show's a feasible way of addressing this. It may not be viable due to binary image size, implications on potential changes to runtimes (which won't work for ROM based RTs which cannot be updated), performance and memory footprint. From a concrete reality perspective, there isn't a single partner who has committed to build the necessary Jco like infrastructure for the embedded ecosystem.

We should be really very clear with everyone about this. In this regard, there is no solution yet. It is important that we call out this gap. When WASI RC 1.0 is released in 2026 what support will P1 runtimes have? Pretty much none - This is critical to know for everyone, particularly in the embedded folks, as it has a direct impact on business continuity.

> > Another critical gap is the absence of low-level, system-oriented APIs that are both efficient and predictable. [...] 
> 
> There are indeed low-level system-oriented APIs that are actively being implemented and proposed in WASI and the Component Model.  While it's true that there is more work to be done to reach the optimal goal state, there's more work to extend WASI Preview 1 as well, and the primary challenge here isn't specific to the Component Model at all; as #7 illustrates, the key tensions are inherent to WebAssembly and concern designing efficient, portable, secure APIs that run across the embedded target matrix this SIG has produced.  Instead, I think we should list the concrete use cases which are currently sub-optimal, under which we can list the solutions being proposed for implementation.

**As noted: ** It's really important to be clear with the reader - these interfaces are currently not present. In this regard the document paints an accurate picture, they don't exist.

I do understand that there are steps to address this and a desire to do so, which is awesome. While there is a lot of good will to try to overcome these issues, I don't recall seeing any contributor promise to provide this, nor do we have a standards process that would mandate it.

So we have to be transparent - right now it doesn't exist. It may not exist in WASI 1.0 RC in 2026.

> > Finally, there is no straightforward way for core WebAssembly modules to interact with Component Model-based binaries. [...]
> 
> There are a number of ways this can work.  One browser-oriented solution is arrived at at the end of [component-model/#275](https://github.com/WebAssembly/component-model/issues/275) which could work analogously outside-the-browser.  Stating "there is no straightforward way" is thus not true; a more accurate statement would be that it is not implemented yet.

**As noted:** Good point. Let's update that to "no currently available way" (done). This now reads:
```
Finally, there is no currently available way for core WebAssembly modules to interact with Component Model-based binaries.
```

> > As a higher-level abstraction, the Component Model introduces inefficiencies and limitations that make it unsuitable for many low-level use cases.
> 
> I don't think this is true as a general statement.  The Component Model already avoids many forms of overhead through its low-level Canonical ABI, and most of the remaining known problems are in the process of being fixed.  Instead, I think we should list the cases where the Component Model's Canonical ABI is currently sub-optimial, and we can then list the particular changes/additions being discussed that would fix these.  As is, the document ignores all the hard work and discussion across this SIG that has been done specifically to address these concerns, such as [CM/#383](https://github.com/WebAssembly/component-model/issues/383), [CM/#175](https://github.com/WebAssembly/component-model/issues/175) and, most recently, discussion around whether we could standardize WAMR's [shared heaps feature](https://github.com/bytecodealliance/wasm-micro-runtime/pull/3823).  Leaving out these developments significantly skews the overall picture.
> 
> > Problem Definition
> > [...]
> > Key challenges include:
> 
> Items 2 and 3 should be reworded to reflect the comments above.
> 
> > WASI 0.2 (previously known as “Preview 2”) – Announced three years ago and officially released one year ago
> 
> WASI 0.2 was released one year ago; I don't know where "announced three years ago" comes from.  The implication seems to be that 0.2 was mostly-done for three years, which is quite untrue; 0.2 only finished implementation a few months before it was released.

**As noted:** There is no implication intended. Only the desire to state the date of the announcement, as this clearly impacts adoption of Wasm / WASI, the traditional case study for this is [The Osborn Effect](https://en.wikipedia.org/wiki/Osborne_effect).

100% let's correct the date. If you've a more accurate date for the announcement, we can change this.

> > WASI 0.3 – Currently in development, this version is expected to be released within the next six months, further iterating on the Component Model-based approach. It introduces enhancements to asynchronous operations, including the addition of future and stream types
> 
> As relevant to this document, we should extend the end of the second sentence with "and a low-level memory-sharing ABI for efficient bulk transfer of data into and out of WebAssembly linear memory".

**As noted:** Could you be a bit more specific? - Are you talking about caller supplied buffers? Or streams (since the sentence states streams already?) - Regarding efficiency, do you have a data point you'd like to share? - I know a lot of the document discusses efficiency, so a data point here would be wonderful, particularly one that contrasts the same operation with native, 0.1, and the 0.3 version. It would be epic!

> > With WASI 1.0 expected to finalize this transition, existing WASI Preview 1-based runtimes and applications face an uncertain path forward. The lack of clear migration options risks fragmenting the ecosystem and leaving many early adopters behind.
> 
> This should be updated according to the above comment to mention the options available to us to allow existing core runtimes to support WASI and the Component Model with minimal additional effort.

Reviewer 2: re: 

> So we have to be transparent - right now it doesn't exist. It may not exist in WASI 1.0 RC in 2026.

I continue to read "may not exist in wasi 1.0" or variations and I want to make it clear that _this is what the spec work is for_, this is why we do this together. Fusion power also may not exist in wasi 1.0, but if we want it to, we make that happen by participating in the work to drive it. That work starts with targets/user stories formally put down. 

I absolutely support and agree that they don't currently exist. That's what this process is trying to create. Per @lukewagner: `Instead, I think we should list the concrete use cases which are currently sub-optimal, under which we can list the solutions being proposed for implementation.` Without concrete use cases listed, we can't design. Without design, no one can implement (and of course who remains an open question that we know about).

re: 
>  I know a lot of the document discusses efficiency, so a data point here would be wonderful, particularly one that contrasts the same operation with native, 0.1, and the 0.3 version. It would be epic!

This is something the SIG can provide by doing the work to test it. Let's make that operational. The SIG should propose to create efficiency targets using the same hardware targets it's already listed, and then propose benchmarking p1 and p3 to see whether they meet the targets and if not by how much. That would be a strong proposal and also actionable.
A document that continues to ask for others to prove something isn't constructive.

Reviewer 2: I note of course that those two items do not address the *transition* from p1 to cm; that's a different bit of work. High order bit is to make cm hit all hardware and efficiency requirements, as if it does not no one will migrate as it provides no business value for the scenarios. Right?

Author: Let’s go back to the context of the document for a moment. We have a situation where many members of the E-SIG have products based on WASI 0.1 or are planning to launch products based on this. WASI 0.1 largely meets their requirements and is via wide market adoption both by runtimes and deployed devices deemed production ready. In contrast WASI 0.2+ doesn’t meet these requirements yet, has one implementation and who’s own creators note it is not production ready.

I can see the comments calling for a contribution to the WASI 0.2+ to make that a better fit for the embedded community. I would, like you very much welcome it. However, the commercial reality I, and many I’ve spoken to the in the E-SIG face make this hard to justify. If we only contributed to WASI 1.0 we would be waiting 2-3 years to have something ready to ship, and stagnation for products already in the field. We can't do this.

There are two key things WASI 0.1 users require, near term additions to WASI 0.1 and long-term support for the 0.1 toolchain. (If I didn’t do a good enough job in the overview, let me know – I’ll happily correct it).

The document, rather diplomatically I think, manages to point out that we in the WASI community have managed to simultaneously employ both the [Osborn](https://en.wikipedia.org/wiki/Osborne_effect) and the [Ratner](https://en.wikipedia.org/wiki/Gerald_Ratner) effect, by messaging that the latest standard isn’t as good as the next (Osborn) and that the only version with wide adoption isn’t good (Ratner) when it actually is – the combined impact is that we are driving people away from both WASI and Wasm. Which is hurting us all.

The document suggests that by providing stability around 0.1, we will arrest the Ratner effect and help with adoption. Helping us all.

>>So we have to be transparent - right now it doesn't exist. It may not exist in WASI 1.0 RC in 2026.

>I continue to read "may not exist in wasi 1.0" or variations and I want to make it clear that this is what the spec work is for, this is why we do this together. Fusion power also may not exist in wasi 1.0, but if we want it to, we make that happen by participating in the work to drive it. That work starts with targets/user stories formally put down.
>I absolutely support and agree that they don't currently exist. That's what this process is trying to create. Per @lukewagner: Instead, I think we should list the concrete use cases which are currently sub-optimal, under which we can list the solutions being proposed for implementation. Without concrete use cases listed, we can't design. Without design, no one can implement (and of course who remains an open question that we know about).
Right. The document attempts to suggest that we could work on 0.1 and share learnings and approaches with folks working on 0.2+. But we have to do this within the resources available, it’s not for a want of desire, it’s for a near term goal and commercial limitation.

>re:

>>I know a lot of the document discusses efficiency, so a data point here would be wonderful, particularly one that contrasts the same operation with native, 0.1, and the 0.3 version. It would be epic!

>This is something the SIG can provide by doing the work to test it. Let's make that operational. The SIG should propose to create efficiency targets using the same hardware targets it's already listed, and then propose benchmarking p1 and p3 to see whether they meet the targets and if not by how much. That would be a strong proposal and also actionable.
A document that continues to ask for others to prove something isn't constructive.

The document is actually full of constructive ideas on how we can address the near-term commercial requirements without hindering the WASI 0.2+ development. We want WASI to succeed.

My remark was a request for clarification from Luke, who has been working hard on bulk data transfer. Luke said it was efficient, so I was asking for the data on that so it could be accurately represented in the document. It’s what we do as engineers and scientists, we look for the data. After all, Einstein not only told us that E=mc2, but he also went on to prove it, it would have made for a very short paper, and lack of publication, dissemination and acceptance if he hadn’t.

Reviewer 2: Oh, my mistake in communication re: `However, the commercial reality I, and many I’ve spoken to the in the E-SIG face make this hard to justify. If we only contributed to WASI 1.0 we would be waiting 2-3 years to have something ready to ship, and stagnation for products already in the field. We can't do this.`

Here I am referring to the work necessary to carry along p1 (`near term additions to WASI 0.1 and long-term support for the 0.1 toolchain`) while the other work carries forward. But this remains an open issue we can figure out once we adopt the precise path.

Reviewer 2: re:
> Luke said it was efficient, so I was asking for the data on that so it could be accurately represented in the document. It’s what we do as engineers and scientists, we look for the data. After all, Einstein not only told us that E=mc2, but he also went on to prove it, it would have made for a very short paper, and lack of publication, dissemination and acceptance if he hadn’t.

I like data; and if @lukewagner can suggest the efficiencies gained now, I wanna hear it, too. I'll just point out that Einstein published a theory that was not demonstrated with data until the last part of it [was confirmed in 2016](https://www.nytimes.com/2016/02/12/science/ligo-gravitational-waves-black-holes-einstein.html). What I mean here is that we either have a story that gives us requirements to hit, or we don't. Is 5% loss efficient enough? Or not? 

Either way, if Luke can make statements here already, I'd love to hear them.

Reviewer 3: > Luke said it was efficient, so I was asking for the data on that so it could be accurately represented in the document.
> ...
> Either way, if Luke can make statements here already, I'd love to hear them.

I said that 0.3.0 makes major improvements to common data-transfer use cases, but I don't think I claimed anywhere that, with 0.3.0, "mission accomplished".  On the contrary, I've been listing the known further improvements we want to make (in some 0.3.x release over the next year).  Anyhow, to avoid the well-known pitfalls of motivated-microbenchmarking (where, if I wanted to, I could create a microbenchmark that is 10x faster on 0.3 than 0.2, but not at all representative), it's probably best to start with a realistic Preview 1 workload that the E-SIG sees regressing with Preview 2 (quite easily measurable, via the preview1-to-2 adapter), and then we can see what happens with Preview 3.  I think that would be a pretty productive starting point.

> If we only contributed to WASI 1.0 we would be waiting 2-3 years to have something ready to ship, and stagnation for products already in the field. We can't do this.

I have full sympathy for this point!  However, while this is a good reason to build and ship something, I don't think it's a sufficient reason to standardize a temporary fork.  (I know this is where the document says "it's not a fork, it's layering", but as I tried to explain in [this comment](https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1980529407), I don't think the layering is technically possible.)

Reviewer 4: > Here I am referring to the work necessary to carry along p1 (`near term additions to WASI 0.1 and long-term support for the 0.1 toolchain`) while the other work carries forward. But this remains an open issue we can figure out once we adopt the precise path.

On that regard, a proposed "path" is the 'layering approach', that tried to cover not only business continuity but also questions around efficiency. @lukewagner pointed out, that this approach is not possible without weakening the nice things of CM and Wasm (sandboxing etc.) (Thx @lukewagner for bringing more light into this).

To my understanding, in terms of "carry along p1 while other work carries forward", the only proposed ways from @lukewagner was to either use (1) WIT incl. canon. ABI for system-level-Calls or bring in (2) adapters (that ultimately - in its runtime performance - also uses the canon. ABI). To my understanding both approaches finally converge to using CM, what is meant to become the basis WASI 1.0. Only (2) covers an ongoing p1 support until every issue with the CM is solved. So, at the end it seems that there is no approach (at least I have not recognized it) to solve the concerns regarding performance. To give some numbers: performance should be near-to-native, i.e. as close as possible to native-speed. A 30% drop of "wasm-aot-compared-to-native" is not close, 5% could be "close enough". Of course, this highly depends on the wasm-workload and the underlying environment.

Now, there has been a lot of discussions around it, where either party wants to see data and the question arises "who is providing those data". I can tell, that this is in general a hard task and numbers on it are also hard to interprete (or to "extrapolate"). So, maybe I should give some examples on what I mean with "performance":

- There are use cases, where a roundtrip of a small amount of data between a wasm module and another native process should be doable in a few 100 CPU cycles. That is, on a given GHz-CPU, a few 100 nanoseconds. Is this doable with _all_ the niceties of wasm and CM? I don't think so. So, this does not seem to be a "smart goal".
- We've already seen, that some workloads seem to put more pressure on cache lines, when using wasm-aot instead of native execution (i.e. more cache misses for wasm-aot for the same workload compared to native). It seems, that there could be some work on doing optimization passes for code generation.
- We've already seen, that e.g. a fourier transoformation on a Cortex-M, having floating-point operations, seem to generate more asm-instructions when coming from wasm (wasm-aot) compared to pure native. This is _maybe_ due some wasm-semantics.
- There are use cases, where the interaction between a wasm module and the underlying system should be doable in single digit us, incl. a bunch of kernel calls. This seems to be achievable.

So, these are just some of the things, that are adressed when thinking about "performance in the embedded domain" and you can see that there is still a lot of work ahead. Because it is hard to come to specific benchmarks, I took a look in where the execution time goes to:
I looked at WAMR and have seen, how the abstraction of a wasi-p1-call (let's say "socket-bind") is done. It is similar to this calling chain: 

> wasm -> wasi-libc-bind -> sandbox-primitive-bind -> os_bind (Posix, FreeRTOS, Windows, ...) -> bind (kernel call of OS; here: Posix-based, might look different on FreeRTOS, ...)

So, imho, seeing this, it is not mandatory to do benchmarks when it comes to judging whether an additional layer helps for "performance". It already takes more calls than just a native implementation, hence there is a performance drop of "wasm compared to native". It is shown, that it is fast enough for realtime use cases on embedded devices (however, I can tell that this is already a close/tight fit for some use cases). But with the CM and the canon. ABI, an additional layer comes on top for such "system calls": this "layer" is the mandatory lifting/lowering part. And this is the thing, that puts concerns on it when it comes to performance. Without doing real benchmarks, it is very likely that performance drops with the canon. ABI for such system calls.

So, regarding performance, we are looking for possibilities to mitigate/omit/optimize lifting/lowering just for at least low-level system-calls. Proposals have been done like "could there be a compiler-switch to assume C-ABI instead of canon. ABI for code generation" or "could a layering approach work"? 
I think, in the embedded domain, there is the will to find a solution that fits. Maybe with some trade-offs, that specific approaches have and are acceptable.

I am happy to dig into this, but most of the embedded domain need the techn. expertise of the champions to help on that.

Reviewer 3: @ttrenner Lots of good points I agree with in your comment.  To push back on one though:

> But with the CM and the canon. ABI, an additional layer comes on top for such "system calls": this "layer" is the mandatory lifting/lowering part.

The goal is that the C-M/CABI is not adding an *additional* layer but, rather, re-rationalizing the *existing* wasm&rarr;host layer, replacing custom ABIs with mechanically-derived ABIs (so that we can derive the benefits of mechanization).  While I know that in the C-M CABI spec that the Python code is literally creating/consuming intermediate Python values, those are just abstract specification-level concepts that are designed to evaporate at AOT wasm&rarr;machine-code time once "both sides" of a call are known.  (Similarly, the core wasm spec reads as-if every wasm instruction takes and produces a GC-allocated array, but we know these evaporate too.)

Now I know that in today's Preview 2/3 implementations, there are some extra calls/layers in the wasm guest code (e.g., in wasi-libc), but iiuc, they are only present due to expediency.  We already want to remove them, it's just a matter of time/resources, so if embedded folks are able to help us here, that would be much appreciated.  Importantly, this is producer-toolchain optimization work that is not dependent on spec/runtime work.

If there are other, more-substantial reasons why extra layers are necessary (vs. what's possible in a custom ABI), then let's add them to [the list](https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1983039221) and figure out what we need to add/change in the CABI to avoid them.

Reviewer 5: > To give some numbers: performance should be near-to-native, i.e. as close as possible to native-speed. A 30% drop of "wasm-aot-compared-to-native" is not close, 5% could be "close enough". Of course, this highly depends on the wasm-workload and the underlying environment.

Do you have realistic benchmarks that show AOT Wasm with WASI p1 are within ~5% of native today?

If so, can you share them? Then we could port them to WASI p2/p3 and have concrete discussions comparing p1 and p2/p3 about actual programs with actual numbers.

If the benchmarks do not exist, then it seems like within ~5% of native performance isn't actually a hard requirement for e-SIG's use cases (because it hasn't even been verified for p1, despite the claim that p1 ["largely meets their requirements"](https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1981875252)) and this is just a synthetic constraint.

Reviewer 6: > To give some numbers: performance should be near-to-native, i.e. as close as possible to native-speed. A 30% drop of "wasm-aot-compared-to-native" is not close, 5% could be "close enough". Of course, this highly depends on the wasm-workload and the underlying environment.

Do you have realistic benchmarks that show AOT Wasm with WASI p1 are within ~5% of native today?

If so, can you share them? Then we could port them to WASI p2/p3 and have concrete discussions comparing p1 and p2/p3 about actual programs with actual numbers.

If the benchmarks do not exist, then it seems like within ~5% of native performance isn't actually a hard requirement for e-SIG's use cases (because it hasn't even been verified for p1, despite the claim that p1 ["largely meets their requirements"](https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1981875252)) and this is just a synthetic constraint.

Reviewer 4: Thx @lukewagner. If I am not way off, then I think there is indeed some misalignment.

You wrote:

> The goal is that the C-M/CABI is not adding an _additional_ layer but, rather, re-rationalizing the _existing_ wasm→host layer, replacing custom ABIs with mechanically-derived ABIs (so that we can derive the benefits of mechanization). 

Your assumption is, that everyone in the world rather would like/can afford to have a mechanically-derived ABI instead of using a custom ABI (usually C ABI). Of course, having it "mechanized" has some tremendous benefits and I fully support that! But I think, that this is not beneficial for every use case out there.
Also, your assumption is, that the existing wasm→host layer is already _always_ doing things like datatype conversions. But, iiuc, this assumption does not hold for every case - they are often just 1:1 mappings. 

Instead, from what I have understood via the discussions in the SIG embedded, using a new ABI instead of the well understood C ABI (or some custom ABI), causes efforts. In addition, often in the embedded domain, most used languages are based on C ABI, hence there is no need to do "additional" data type conversions if the embedding host is written in e.g. C. And even if there are differences: often they can be overcome by doing some "ugly/error-prone but fast re-interpretation" without doing lifting/lowering using a newly created generic/canon. ABI.

In that sense, being **_forced_** by the _future_ toolchain to use the lifting/lowering instead of a custom ABI is an additional layer compared to what is today used broadly in the embedded space. I imagine, that if the toolchain (compilers, sdks, in combination with runtime capabilities etc.) gives the option to not use lifting/lowering (and wit offers the option to e.g. "annotate" that) and instead uses either C ABI or existing approaches, then this would be seen to be helpful.

I admit, that such a task imposes efforts and it would be necessary to sketch out all the pros and cons, and how the burdens are to be shared (even if the current adopted wasi-p1 approach is already supported sufficiently by wasi-sdk and runtimes).

Please give me a hint, if I have some misunderstandings on my side.

Reviewer 4: @fitzgen 

> Do you have realistic benchmarks that show AOT Wasm with WASI p1 are within ~5% of native today?

The honest answer is easy: "No, I don't have _real world_ benchmarks". I cannot go too deep in explaining why this is the case, but it is connected with the fact of having different personas that interact with our envisioned system.

However, of course I have "synthetic" benchmarks. A purely big arithmetic one creates - unoptimized - hundreds of thousands of stack variables. So, using wamrc, we optimize it and it boils down to a double digit number of stack variables. Using such an arithmetic benchmark, running on a x86 and a Cortex-A CPU showed, that wasm-aot is truly near-to-native-speed with native execution (which is not a big surprise, when you think about it). I am pretty sure, that also with the C-M, there is not a big difference.

We have other benchmarks, that where already published by @woodsmc (I guess back in 2022), that showed, that execution with wamr was faster than with wasmtime (with permuted, but comparable optimization settings and compilers (llvm/cranelift)).

And we have some internal benchmarks, running on a Cortex-M (~120MHz, FreeRTOS), that does a fourier-transformation (you can get some as open source) by transferring an array with 1024 complex numbers consisting of two floats can cause a drop of 100% using aot (hence, we must have a closer look into it to understand why that is the case and what can be done to improve it). That benchmark _might_ have no meaning regarding C-M or not, but we do not know, yet. Note, that we were not able to do that test with wasmtime, because we were not able to port wasmtime to FreeRTOS in a short time, while this was possible using WAMR.

All in all, our findings match with benchmarks that also have been done by various numbers of persons/stakeholders and which are already published (one example [here](https://00f.net/2023/01/04/webassembly-benchmark-2023/)).

Regarding your statement that "~5% to native" is not a hard requirement and a synthetic constraint:

> If the benchmarks do not exist, then it seems like within ~5% of native performance isn't actually a hard requirement for e-SIG's use cases (because it hasn't even been verified for p1, despite the claim that p1 ["largely meets their requirements"](https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1981875252)) and this is just a synthetic constraint.

I can tell, that the constraint of getting it as fast as possible is a very real constraint, as we are talking of realtime systems with guaranteed reaction times (i.e.: signal from a sensor -> calculating the reaction -> setting an actor).
So, on a performance manner, we don't have much room for maneuver - it is tight even with pure native execution. But the benefits of wasm are undoubtedly worth to very seriously consider it.
That said: _any_ optimization regarding performance is highly welcomed, independent of the concrete use case, while _any unnecessary_ additional indirection should be avoided.

Reviewer 3: @ttrenner 

> Your assumption is, that everyone in the world rather would like/can afford to have a mechanically-derived ABI instead of using a custom ABI (usually C ABI). [...] But I think, that this is not beneficial for every use case out there.

You're right that there are special cases that require special support, and in such cases we have several options with the mechanized approach: add new built-ins (with custom core signatures), add new WIT-level types that have more-advanced ABIs, add more-advanced ABI *options* to change the ABI of existing types, or some combination of these.  This is already what we're doing with 0.3.0, as well as the follow-on additions that I've been mentioning already in this PR.  If we expected 100s of such additions, this approach of course is too onerous.  But pragmatically, I don't think there are.  Again, I think enumerating concrete use cases *in WSI* in a list and talking through each one in detail could help us feel more at ease.

Author: I think what we are talking about here is the disproportionately larger impact that additional instructions have on lower powered (embedded) processors compared to server grade hardware, particularly in relation to control loops. I don't think benchmarks would necessarily help when conveying the differing perspectives. Perhaps if we thought about this from the number of CPU instructions executed and latency.

If we consider a function which needs to pass a buffer of *n* bytes, if we pass this via a pointer, it is of course O(1). If we have to copy this data it would be O(*n*). 

On cyber physical systems we are largely I/O bound, reading external real world state and instructing external real world devices to move. Given the processors speed and the reaction time we must have, there is a budget of instructions and clock cycles we can spend consuming, processing and responding to the outside world.

Available Instructions for Logic = (required response time * instructions per second) - (instructions for input data + instructions for response data)

If we take one of the processors on the E-SIG list, let's say the ESP32-C6, Risc-V processor. It has a clock speed of 160Mhz, so roughly 160M instructions per second. Let's assume we need to reply to an input signal in say 20 microseconds. That would be ( 0.00002 * 160M) = 3,200 instructions. (note: Microseconds, for some use cases can be really big - occasionally we deal with single digit nanoseconds). 

Now we must remove from that 3,200 the number of instructions used to read the state of the outside world and output our response. If the cost of data input / output is O(1) then we can take a fixed cost away 3,200 - 2*O(1). If the cost of data input / output is O(*n*) then of course we have 3,200 - 2*O(*n*). 

I know the cost of using a CM can be mitigated by putting everything into the component, but if a component barrier exists between the code and the host, then I still face this input / output cost.

Now, outside of the control loop, where we may interact with an enterprise application, then yes, that's where the component model will offer huge advantages. But for this tight timing it's impact can be pretty large.

As an aside - It also means that the async like interfaces being proposed are not usable for us. Since we'd need tighter control of the scheduling and timing to meet the latency and response times we must deal with.

[Edit: Typo]

Author: > @ttrenner
> 
> > Your assumption is, that everyone in the world rather would like/can afford to have a mechanically-derived ABI instead of using a custom ABI (usually C ABI). [...] But I think, that this is not beneficial for every use case out there.
> 
> You're right that there are special cases that require special support, and in such cases we have several options with the mechanized approach: add new built-ins (with custom core signatures), add new WIT-level types that have more-advanced ABIs, add more-advanced ABI _options_ to change the ABI of existing types, or some combination of these. This is already what we're doing with 0.3.0, as well as the follow-on additions that I've been mentioning already in this PR. If we expected 100s of such additions, this approach of course is too onerous. But pragmatically, I don't think there are. Again, I think enumerating concrete use cases _in WSI_ in a list and talking through each one in detail could help us feel more at ease.

This looks like a great idea for a more detailed discussion around a second more focused document - https://github.com/bytecodealliance/sig-embedded/issues/20 ?

I'd love to get this current document merged, with some edits around language. It has done it's job in provoking the discussion. If we move on to some extra detail and some more concrete actions that would be awesome.

I think for the next SG meeting we should present on the use case and need for P1 support... essentially setting out the requirements for the discussion, and outlining what the goals of this discussion would be.

Reviewer 6: @ttrenner 

> We have other benchmarks, that where already published by @woodsmc (I guess back in 2022), that showed, that execution with wamr was faster than with wasmtime (with permuted, but comparable optimization settings and compilers (llvm/cranelift)).

We are not discussing engines, we are discussing APIs to external resources and the potential overheads of language features, so that benchmark comparing WAMR and Wasmtime isn't relevant at all.

What would be useful would be taking the same program(s) -- ideally one(s) the SIG cares about or believes to be realistic/representative -- and compiling it(them) to both `wasm32-wasip1` and `wasm32-wasip{2,3}` and running both versions in the *same* engine so that the only variable changing is which version of the external APIs are being used. That would give us information about what kind of overheads, if any, the newer versions of the APIs impose for these programs that we care about.

> And we have some internal benchmarks, running on a Cortex-M (~120MHz, FreeRTOS), that does a fourier-transformation

Based on your description, this sounds like a purely compute workload that minimally interacts with the filesystem or other external resources, and therefore not useful for evaluating APIs to external resources. It seems like it would be fine for comparing Wasm execution throughput between engines or to native (modulo the usual qualifications that micro benchmarks are not representative) but it is not useful for this discussion about external APIs.

Reviewer 4: @fitzgen Yes, you're right. My above statements where just about to highlight which "dimensions regarding performance" we are looking at.

I'll take your suggestion to discuss that in the SIG embedded.

Beyond the WASI standard, there are academic and experimental implementations which have explored extending WASI Preview 1\.  Projects such as the WebAssembly Linux Interface (WALI)[^1], which explores exposing the Linux kernel APIs via WASI core wasm function imports and exports, paving the way for WASI interfaces to be implemented in a portable, layered way have also been introduced and evaluated.

## Long term Toolchain Support \- Continuing the Legacy of Success

A stable and long-term supported toolchain is essential for any healthy platform and WASI is no exception. Toolchains form the foundation for application portability, enabling developers to build and maintain their software.  A stable toolchain gives developers confidence that their binaries will remain functional across different runtimes and will continue as WASI evolves. However, the ongoing evolution of WASI, particularly the shift introduced with WASI 0.2+, has raised concerns about long-term stability.

> The meaningful use of WASI Preview 1 (P1) means that many runtimes and applications depend on its established behavior. Developers have invested significant time and effort into building applications, frameworks, and infrastructure around WASI P1, making it a critical part of the WebAssembly ecosystem. Abandoning these early adopters without a viable migration path risks fragmenting the ecosystem and undermining trust in WASI’s long-term viability. Instead, fostering a healthy ecosystem requires maintaining continuity—allowing developers to leverage improvements in newer WASI versions while ensuring their existing investments remain functional.

Reviewer: > Abandoning these early adopters without a viable migration path...

This is not a fair judgment of how toolchains like wasi-sdk and wasi-libc have approached the requests for longer support of p1 &mdash; I would say this language is a mis-characterization and should be fixed. I have had many conversations with @woodsmc, @loganek, etc. in which they expressed a desire for long-term support for p1 in those toolchains. And, in fact, here's one of my efforts to support that desire from exactly one year ago: https://github.com/WebAssembly/wasi-libc/pull/476.

But note the lack of feedback from the "keep p1 alive" parties in that PR: I believe that it is the responsibility of those who want to maintain p1 stability to be invested in that work, not leave it to others. Those who are moving the specification forward (p2, p3, etc.) should not be forced to have the same stability motivation as the "keep p1 alive" parties. And that is an open invitation for the "keep p1 alive" parties to increase their support for the parts they care about; I feel both goals, progress and stability, are still clearly possible at this point.

Reviewer 7: Maybe it's an oversimplification / mischaracterization but I believe the largest concerns with preserving p1 interfaces is more about preserving core wasm module format and interfaces. I'm optimistic that https://github.com/WebAssembly/component-model/pull/378 will provide the alternative that the "keep p1 alive" parties would want without the current limitations of what is in the scope of p1 / extending the interfaces in p1.

Reviewer: @rajsite, yeah, I agree with your "largest concerns" assessment. And I guess that strengthens my point: it is not as if toolchain maintainers have "abandoned" p1, but rather that the companies represented by this document should invest in making the issue you link to (and/or my draft PR) a reality.

Author: Hey @abrown, thanks for considered thoughtful feedback, it really is much appreciated. In large what I hear you saying actually aligns with the document. You are highlighting the need for those organizations who require the 0.1 toolchain to get involved with it, as the those  taking the standard forward don’t see a need to do so.

There is no judgement in the document. But the document does need to very clearly and very transparently describe the state of current support for the WASI 0.1 toolchain, along with the wider implications of this.

There are organizations I have met who believe that WASI 0.1 is going to have indefinite support. They believe that this support will be provided for by those moving the standard forward. This is not my understanding. Reading your comment you seem to suggest the same thing. The mistaken believe that there will be indefinite support has resulted in a lack of attention on WASI 0.1 stability – something the document is trying to address.

Are there any published plans for WASI 0.1 support, that the document authors may have missed? – If so then yes then we should accurately reflect these in the document.

My honest, personal understanding is that this does not exist. My understanding is driven by my personal conversations with you, other members of the BCA TSC and folks in Fastly and Cosmonic. When I enquired about longer term support the response I received was the same from every party, that WASI 0.1 was a preview and should not be used. As I explained on the E-SIG call, and included in the overview of the document, that boat has sailed. We (those people using WASI 0.1) need to provide this missing longer term support.

As we move forward to provide this support and start defining the concrete next steps there may be initiatives that we’ve not been involved in which we should. The PR you linked to is a great example of that, and I’m grateful you pointed that out.

Regarding the next steps, I am much more interested in yours and others’ opinions on how the renewed investment in WASI 0.1 can be managed alongside the evolving 0.2+ releases. The document authors spent a considerable amount of time trying to work out a compromise that could work for all.

I can see the WASI 0.1 / Embedded folks wanting to move forward with some new additions to WASI 0.1 to provide missing system level functionality. I can envision this happening within the next month or so.

Author: > @rajsite, yeah, I agree with your "largest concerns" assessment. And I guess that strengthens my point: it is not as if toolchain maintainers have "abandoned" p1, but rather that the companies represented by this document should invest in making the issue you link to (and/or my draft PR) a reality.

This appears to be a call to action. I love this, and I really appreciate you calling out the need for investment. It is something we should consider in the E-SIG as we define the concrete next steps.

Reviewer: > In large what I hear you saying actually aligns with the document [...]  But the document does need to very clearly and very transparently describe the state of current support for the WASI 0.1 toolchain, along with the wider implications of this.

I do understand many of the underlying concerns but what I'm pointing out is that what the document is describing, whether clearly or transparently, is actually incorrect: IMHO, p1 users have _not_ been abandoned by toolchains and there _is_ a viable migration path (see the links above).

Reviewer 4: From what I've seen it'd be correct to state that there are existing approaches to support migration in terms of the generation of wasm modules (by wasi-sdk) (based on your sketched [approach](https://github.com/abrown/wasi-libc/blob/preview2-plan/docs/preview2-transition.md)) and proposals that _could_ help to make p2-modules workable on p1-runtimes (see [Marcin's proposal](https://github.com/loganek/wasi-snapshot-preview2-to-preview1-adapter)).

Other than that, the document tries to state that _there are_ gaps in current p2 tooling that reflect p1 functionalities and non-functional requirements of the embedded domain. That does not mean, that there are no ideas of how to do that based on p2. _It just means, that it does not exist yet_. Plus, there are concerns that - in case that there will be no means/support for low-level api support - P2 _might_ not do the trick for the needs of the embedded domain. This concern might be wrong, but it is reality, that p1-based products are already running on tons of "embedded" devices. So, p1 is "proven in use" and hence the risk of extending and improving p1 _seems_ to be lower than hoping that P2+ will work for the embedded domain in future.
The concerns about P2+ could somehow be tackled if there is an agreement to place quality gates in the phases. Note, that this is also brought up in the document. I have not seen any comment on that, which makes me feel like that there is little to none support for that. Again, bear with me if I am wrong with that - maybe there was no opportunity to dig into this.

The document also tries to sketch an solution to let both concepts (P1-like and P2+) work: a layered approach. The bottom layer is just about supporting low-level APIs, while the top layer does all the nice things to get interlanguage interoperability (component model) - by the way: having a well-defined way for FFI declaration _is_ wanted by the embedded domain but not for the price of performance, when performance is more important and Posix, using C ABI etc., does the trick (also accepting the trade-off of being forced to accept another way for FFI declaration or the trade-off of not having the possibility to mix languages).
Now, there are comments that the layering approach might not be necessary, because it might be doable with P2+ approaches. This seems to be _a bet_ in the future, at least as long as there are no working examples showcasing this broadly. I admit that there are examples when looking at shared memory. _Stream_ of P2+ seems to serve well for a couple of use cases. But _there are_ use cases where stream does not seem to be the best solution (e.g. in case random access is needed and well-known MMU features (using mmap) are enough for a specific use case). Plus, there are different approaches out there, e.g. WAMR's shared heap (which also does not fit for all the uses cases that need "shared memory": e.g. if sharing is needed across different processes (albeit with all the cons that come with it!)). 

There is legacy software, that _is_ based on Posix. There is legacy software, that should be ported easily to the wasm world, i.e. clone the repo and issue clang/cmake and generate a wasm module. There are missing gaps like signaling (which has been discussed [here](https://github.com/WebAssembly/WASI/issues/166) back in the days). In the linked issue it is highlighted, that signal handling is not an easy task and maybe not everything is needed (and not all of it might be sensible in the "Wasm realm"). Also, maybe, not everything of Posix is needed. But we do not know yet. But we know that there are examples that need to set thread-attributes or socket options (e.g. unblocking source). Such software needs more "posix like" support and it feels to become overcomplicated (incl. performane loss) to map (and re-do?) it to p2+ approaches if it is not needed for a given use case.

Long story short: Anticipating all the different use cases does not seem to be realistic and hence we are advocating for a more flexible approach of interacting with the "underlying system" - I think that the layered approach is worth to be discussed. I'd be irritated if this is not even considered worth to talking about.

[edit] changed some wording and typos - sorry for the inconvenience.

Author: > > In large what I hear you saying actually aligns with the document [...]  But the document does need to very clearly and very transparently describe the state of current support for the WASI 0.1 toolchain, along with the wider implications of this.
> 
> I do understand many of the underlying concerns but what I'm pointing out is that what the document is describing, whether clearly or transparently, is actually incorrect: IMHO, p1 users have _not_ been abandoned by toolchains and there _is_ a viable migration path (see the links above).

@abrown - There has been a tremendous amount of engineering and effort expended to keep the WASI 0.1 tool chain functioning as it is today. You are right we should call this out in the document. It is something I know all users of WASI are very grateful for. 

The document is doing a little bit of a hamfisted job of it, but it attempts to point out that there is such a resounding lack of commitment from anyone, either the users of 0.1 or the contributors to moving the specification forward to maintain that support. We know that there are commitments to continue work on WASI 1.0+, but nothing exists for 0.1, this lack of support (it doesn't appear deliberate, more like it's fallen between the gaps) is effectively the same as abandoning the 0.1 users. This is particularly true when you combine it with the lack of WASI 0.1 runtimes to execute code targeted at WASI 0.2,0.31.0RC and beyond. The result is a tool chain that's no longer supported and a runtime that can't execute the latest binaries... 

In drawing attention to this, the intent in the document, is to demonstrate the need to step up and contribute.

I'd appreciate some help with the wordsmithing to get this all in the document.

Reviewer 2: Yes. Two concepts: 
1. is there a technical path toward useful support until components deliver the same feature set on target devices?
2. who will implement and maintain that technical path?

Those concepts are both required in the new document, methinks.

Reviewer: In response to @ttrenner, I would like to point out that I am quite sympathetic to the desire to avoid unnecessary overhead when accessing HW; I'm sure I've annoyed some with my intent focus on this in the past! So I am all in favor in discussing ways to make things more performant and this document is aiming towards that direction &mdash; all good! What you may be missing is that I'm objecting to is the blame-based language used in the document, which @woodsmc picks up on in his comment. It feels a bit unfair to those of us who do work on these toolchains and have tried, as time permits, to help. My opinion here is that current p1 support and the migration to p2 has not been "abandoned" as much as "not worked on," hence my call to get more implementations involved; like @squillace points out, "who will implement and maintain?" And @woodsmc, I'm not sure I agree with your latest conclusion that "the result is a tool chain that's no longer supported," &mdash; do you mean this is the state you see today or a state that could be some future? My take is `wasip1` is still being used effectively out in the wild for several languages...

I don't want to put words into your mouth, but to wordsmith this, could we:
- retain the concerns about toolchain gaps that may appear in the future; sure, more _must_ be done if long-term stability is a goal
- link to some of the examples I mention as plans that would improve the situation but require implementation
- make a call to implement in that direction (it doesn't have to be those specific ideas); I feel the real focus of these paragraphs should be just that, motivating the implementation and maintenance of high-quality, stable previews
- avoid blaming the ecosystem that the eSIG is benefitting from; perhaps some recognition for the engineers who have brought the toolchains to this point (and I don't mean me, there are many who have contributed much more here!)

@woodsmc, what do you think of that?

Author: @abrown Thank you! - and hey, don't worry about putting words in my mouth at all - It was a honest call for help, and I really appreciate your input! - so, yeah, big thank you.

I agree. Let's go through the document and make sure it doesn't *blame*, that's not the intention. The intention is to detail where we are the moment in a kinda open and frank way. I've unfortunately not been able to hit my own milestone, and I apologize for that.

What I can see, (if I attempt to TLDR this in a comment), is that there has been an initial investment in WASI which has been tremendously successful. The new investment (P2+) is heading in a direction which will not serve a not-insubstantial subset of the users of the original WASI work, mainly members of the E-SIG, and there is now a need to address this. The direction of the new investment also leaves those who invested in 0.1 runtimes without a migration path to execute the output of future WASI SDKs.

There is no blame here. It "just is what it is", or "we are where we are". But in addressing it this way, and detailing what would happen if it is not addressed, the document attempts to get the reader focused on the need to contribute.

I also think we need to highlight where some of the investment is going to need to come from to address this. The document attempts to point out, that the need and desire of those organizations who are driving WASI at the moment, to maintain the WASI 0.1 SDK declines as WASI 1.0 approaches. - this isn't wrong... it makes sense for those organizations... no harm, no fowl, it just is what it is.

Therefore... those folks using the WASI 0.1 SDK need to start getting involved and helping to contribute.

This is mainly covered in the first half of the document. The second half covers how this might be done.

I'd like to update the document to ensure that the tone is right, but to also make sure that the need for additional contributions is still there. I will go through this on Monday and update it.

I also agree with Oscar's feedback, that we can do two derivative documents, one which details the need to get involved and the WASI 0.1 SDK need for the E-SIG, and the second for talking more about how we'd do that.

I feel like this document has really done it's job of provoking the conversation and as an internal E-SIG discussion piece it's been pretty good in doing that. Now's the time to move on to more concrete steps, how do we support the 0,1 SDK? and if we add to it, how do we do that?

By clearly defining these layers, we can allow high-level APIs to be built on top of low-level primitives without imposing unnecessary constraints. The WebAssembly Application Interface can provide flexible, structured APIs built around the Component Model and the Canonical ABI.  However, these application-level interfaces should be distinct from the low-level WebAssembly System Interface, which provides fundamental operations that are difficult or impossible to emulate.

### Two Related Standards

> To support this layered approach, we propose splitting WASI into two separate standards, each with its own scope and governance:

Reviewer 3: I'm afraid I don't see how what is described here can be called "layering" when it comes to *standards*.  In particular, the "WSI" interfaces would have custom-designed ABIs that are unrelated to WIT and the Canonical ABI.  Thus, if the "WAI" layer wanted to use the "WSI" layer, someone would have to hand-write an adapter for each WSI function to the Canonical ABI (in the same way that folks had to hand-write the Preview1-to-2 adapter -- a tricky bit of work).  Thus, this is not a layering of standards.  By the same token, one can implement WASI in terms of Web APIs, but the WASI standard is not "layered" on Web standards.

If there is no "layering" relationship, I don't think it makes sense to attempt to introduce "WSI" as a separate standard as part of WASI.  This will make an already rather-complex space of standards even harder to understand since there is no "factoring out" of common concepts; it's just more moving pieces.

Now, I do appreciate that, until we implement improvements to the Canonical ABI (caller-supplied buffers, removing cabi_realloc, etc), embedded folks wanting to ship something in the short term (which, iiuc, is another implicit goal here) may not want to use the Canonical ABI -- there are valid reasons to want to hand-craft new ABIs in the interim.  But we also don't need to standardize these interim things for them to solve our short-term needs (indeed, attempting to actually standardize them will just work against the goals of being *short term*).

Thus, I'd propose that the E-SIG reframe "WSI" to be a WAMR project that extends Preview 1 with additional functionality but also with an eye toward convergence with the Component Model along with the explicit understanding that, to converge, folks in the E-SIG will need to hand-write an adapter between WSI and the Canonical ABI (analogous to the preview1-to-2 adapter).  As long as the E-SIG provides the engineering resources, I can see this being a valuable addition and I'd be happy to continue working with folks in the E-SIG to plan the convergence, enumerate the remaining gaps and close them.

Reviewer 8: The concept of layering here is about separating lower-level and higher-level abstractions in a way that aligns with our standards approach. Rather than conflicting with it, this proposal fits within the existing process and maintains alignment. 

I think this diagram shows illustrates this concept clearly:
![wasi-layered](https://github.com/user-attachments/assets/98e2dc0b-ca75-42cd-b73d-975cd722ac7d)

At the lower level, WSI would be limited to Core Wasm constructs, much like a minimal kernel interface. This keeps the implementation lightweight and ensures minimal performance overhead—critical for systems and embedded programming. The API surface would be small and map closely to functionality present in every major OS today.

At the higher level, the Component Model provides full expressivity, including rich types, async calling, a fully shared-nothing model, streams, futures, and more. Splitting these concerns allows for a cleaner design: for example, a low-level BSD-style socket API at the WSI later could coexist with a fully async `wasi::sockets` interface, similar to how Java and other ecosystems offer both low- and high-level abstractions.  This enables developers to choose which option best suits their needs.

Regarding hand-crafted adapters, I agree that we should avoid them. The real issue is that WIT currently lacks the ability to represent Core Wasm imports and exports—something that is ultimately a tooling gap we can address.   

As for reframing WSI as a WAMR project, I don’t believe that’s the right approach.  I think that would presents a greater potential for fragmentation or a fork which I think we all wish to avoid.  

Finally, on the topic of engineering resources, the engagement within e-SIG suggests strong interest in this approach. While engineering contributions aren’t guaranteed, the enthusiasm we've seen is encouraging. Moving forward will require work on tooling, interfaces, and SDKs (particularly WASI-LIBC), but the potential benefits make this a worthwhile direction.

Author: Hey @lukewagner, @srberard - as @ospencer suggested this looks like a standards discussion, so we should move this to the SG.

Let's get this document merged, then prepare the simplified use case justification, and the following forward compatibility / roadmap technical note.

I think that would probably be the best way to expand upon this discussion, and as @ospencer reminded us, move it to the right venue.

It is a really interesting discussion, it would be interesting to see what others think about this too.

Reviewer 3: > The real issue is that WIT currently lacks the ability to represent Core Wasm imports and exports—something that is ultimately a tooling gap we can address.

I don't think that's true; we should really try to bottom out on this since I think it's at the root of a lot of the misunderstanding here.

To provide some background: there are only two kinds of imports provided by the Component Model's Canonical ABI to core wasm:
* **WIT-derived functions**: core function imports that are derived mechanically from WIT function signatures via the Canonical ABI (this is probably what you're thinking about)
* **Built-in functions**: core function imports that are defined directly *as core functions* by the Canonical ABI with no associated WIT and that do not need to be imported by the outer component

Built-in functions give us the ability to craft custom ABIs and indeed that's what we're doing in 0.3 with the `stream.read` and `stream.write` built-ins, which take a pointer to caller-supplied memory that is held onto.  I expect we'll add more built-ins for [caller-supplied buffers](https://github.com/WebAssembly/component-model/issues/175) and perhaps WAMR-style shared heaps.

However, we have to be *very careful* when adding built-in functions because they can't be sandboxed or virtualized; they're like wasm instructions (like `memory.grow`). Thus, if you look at the current [list of built-ins](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#canonical-built-ins), they adhere to the same [capability-safety](https://en.wikipedia.org/wiki/Capability-based_security) rules that Core WebAssembly adheres to by falling into one of 3 categories:
* They create something new and isolated (like core wasm `memory` or `struct.new`)
* They access something given to them via unforgeable handle (like core wasm `struct.get` or `i32.load`)
* They do something ambient that has no security implication (like core wasm `atomic.fence`)

Of course most I/O (filesystems, sockets, keyvalue, http, gpu, i2c, usb, ...) doesn't meet this criteria, and so for all these we need to use WIT interfaces that are explicitly imported by the component and are virtualizable.  (And with the current and future set of built-ins, I think these WIT-defined interfaces can be sufficiently performant for all our embedded needs.)

By following these rules, we can get a lot of tangible benefits that are important to the folks working on the component model: sandboxing (supply-chain-attack resistance, principle of least authority, ...), safe host bindings, emulation for local testing, record-and-replay debugging, and bunch more automated tools.

When I look at the list of WSI interfaces in the above diagram, most of them do not meet the criteria for being built-ins.  Thus, we can't simply say "the Component Model can treat all the WSI functions as built-in functions", because they break the key invariants we want from the Component Model.  So if they are not built-in functions, they need to be WIT-derived functions, and if they don't use the Canonical ABI, then we'll need an adapter.

> Regarding hand-crafted adapters, I agree that we should avoid them.

Hopefully the above explains why I think that either (1) we use the Canonical ABI for the WSI interfaces that can't be built-ins (which is most of them) or (2) we need an adapter.  I'm not aware of a third option that doesn't break all the useful Component Model properties, but let me know if I'm missing something.

Again, if folks have short-term needs that can't wait for the Canonical ABI to improve, I think there's no problem designing an interim custom core wasm interface in WAMR that later we have to adapt to the Canonical ABI, but I don't think we should go through the expensive process of trying to standardize that if the goal is "short-term".

-----

There's one other independent argument for why the Component Model can't be layered on WSI: native browser implementation.  This is an important goal but browsers have clearly indicated that they are *only* open to natively implementing the Component Model (b/c, as I explained above, it adheres to the same rules as core wasm and doesn't "bake in I/O").  In contrast, there is little chance browsers would natively implement WSI, so if the Component Model were layered on WSI, that would knock out the Component Model.  This is why WASI is layered on the Component Model, and not vice versa -- browsers can implement the Component Model natively (including the Built-in Functions, which would need to be implemented natively too) without implementing WASI (WASI would be polyfilled, even quite efficiently if we have a native Component Model implementation).

Reviewer 3: @woodsmc (Sorry, I didn't see your comment when replying) Discussing in the SG too sounds fine.  Before merging, it would be good to see the basic comments in #10 fixed; as is, there haven't been any iterations since the original draft.

Author: @lukewagner - No worries. Thanks for your patience while I switched from the issue to the PR structure. Following @abrown reminder, I copied the issue and inlined the feedback in the comment above - https://github.com/bytecodealliance/sig-embedded/pull/18#discussion_r1972763787

I know it is easy to miss, but there have been two edits, and a request for clarification. I think the linked version is a bit easier on the eye than the original issue. Happy to resume the conversation from there.

Reviewer 8: @lukewagner - Thank you for your explanation regarding what is available today with WIT.  I agree that built-in functions need to be carefully constrained to maintain the security and virtualization properties of the Component Model. I’m not advocating that WSI functions be built-ins; I completely agree with your reasoning on that point.  WSI functions should be defined formally in something like WIT.  WITX offered this but had a number of shortcomings.  Extending WIT to support expressing Core Wasm imports/exports would be a way to address this and something I would be open to.  Again, the proposal is for WSI to be limited to Core Wasm only; hence these would need to be limited to Core Wasm constructs (and thus Core modules imports/exports) 

Regarding browser implementations, I don’t see how WSI, expressed as Core Wasm imports/exports, would prevent browser implementations.  Couldn’t a browser provide these interfaces directly or through an adapter, similar to other host functions.

+1 on discussing this in the SG

Reviewer 3: > Extending WIT to support expressing Core Wasm imports/exports would be a way to address this and something I would be open to

Ah, this may be the key point: unfortunately only built-ins can have raw core wasm signatures.  If we tried to add core function signatures to WIT, the problem we hit (concretely, already with WITX) is that almost all core functions depend on there being a single ambient shared memory and function-table.  Once you depend on ambient shared memory/tables, the useful properties of component model are lost (security sandboxing, low-cost virtualization, simple automatic bindings, etc).  By improving the Canonical ABI and the set of supporting built-ins&mdash;addressing performance cases where you'd otherwise want a core function in WIT&mdash;I don't think we'll *need* core functions in WIT.  This is what I'd like to enumerate in a document and work through, because while there is a list, it's not a huge list.

> Regarding browser implementations, I don’t see how WSI, expressed as Core Wasm imports/exports, would prevent browser implementations. Couldn’t a browser provide these interfaces directly or through an adapter, similar to other host functions.

Oops, sorry, I should have contextualized this by saying that this is a problem if WSI interfaces are to be treated **as built-ins** (due to their use of core function signatures that depend on sharing memory/tables).

Reviewer 4: @lukewagner @srberard I think, we come closer to more concrete action items. 

> If we tried to add core function signatures to WIT, the problem we hit (concretely, already with WITX) is that almost all core functions depend on there being a single ambient shared memory and function-table. Once you depend on ambient shared memory/tables, the useful properties of component model are lost (security sandboxing, low-cost virtualization, simple automatic bindings, etc).

> I don't think we'll need core functions in WIT

I think, that in addition to saying that this and that is not working (due to cannibalizing CM's niceties) it would be good to sketch a concrete _generally working proposal_. I think, that it is near to impossible to think of all the future use cases that might come up. Just taking "stream" as an example: it adresses use cases for which sequential data exchange is sufficient. It seems to not be sufficient for random access data exchange. Another example, not dependent on CM yet: WAMR's shared heap: It addresses use cases, where "shared mem between wasm modules is needed", but does not address use cases where shared mem between wasm and another native process is needed.

So, a generally working concept would be great. I must admit, that I don't know whether this is possible at all. But you, @lukewagner, seem to be optimistic. So, maybe we as a community could - together - work out, how a concrete solution for optimizing performance with WIT/canon.ABI might look like. And once all agree on that general concept, we could try to share the burden of implementing it.

Reviewer 9: >  Another example, not dependent on CM yet: WAMR's shared heap: It addresses use cases, where "shared mem between wasm modules is needed", but does not address use cases where shared mem between wasm and another native process is needed.

Apologies for any confusion caused. [The original concept behind the shared heap](https://github.com/WebAssembly/memory-control/issues/19#issuecomment-2537855162) was to allow a Wasm application to access a specific part (or the entirety, if trusted) of the host memory that lies outside the sandbox and may even be in different address spaces. It's somewhat akin to mapping a portion of the host memory into the Wasm world, thereby extending the linear memory defined by WebAssembly. 

From another perspective, it can be seen as additional linear memories created by the host(like `(memory $2)`, `(memory $3)`). If there were a usable toolchain that supported the multi-memory proposal, this shared-heap feature would be unnecessary and might not have been developed. Personally, I see it as a contingency for when there is a lack of spec support, and a product will find its own solution.

Reviewer 3: @ttrenner That sounds like a great next thing to work on together as a SIG.  I'd be happy to iterate with everyone on such a goal-state document that addresses the use cases we have in mind.  And you're right: 0.3 streams don't solve all the use cases; more is needed, so a great next step is to enumerate the remaining specific use cases and which specific additions (to WIT/CABI) would address them.

## Long term Toolchain Support \- Continuing the Legacy of Success

A stable and long-term supported toolchain is essential for any healthy platform and WASI is no exception. Toolchains form the foundation for application portability, enabling developers to build and maintain their software.  A stable toolchain gives developers confidence that their binaries will remain functional across different runtimes and will continue as WASI evolves. However, the ongoing evolution of WASI, particularly the shift introduced with WASI 0.2+, has raised concerns about long-term stability. For many organizations, particularly those maintaining long-lived software, frequent runtime changes are not feasible.

The meaningful, widespread use of WASI Preview 1 (P1) means that many runtimes and applications depend on its established behavior. Developers have invested significant time and effort into building applications, frameworks, and infrastructure around WASI P1, making it a critical part of the WebAssembly ecosystem. 

> There has been considerable effort expended to ensure that the WASI 0.1 tool chain remains useable while the WASI 0.2+ transition continues, an example of this is the [WASI Preview Transition](https://github.com/abrown/wasi-libc/blob/preview2-plan/docs/preview2-transition.md). There is as yet no way to run WASI 0.2+ binaries on a WASI 0.1 runtime, although there have been some investigations, like [this one from Marcin](https://github.com/loganek/wasi-snapshot-preview2-to-preview1-adapter), non have been successfully reached a conclusion and been used in production. Additionally, the long term continued investment in WASI 0.1 is not guaranteed. The organizations supporting WASI 0.1 are the same organizations driving WASI 0.2+, as WASI moves towards a 1.0 release these organizations will not have the need to continue to offer support for WASI 0.1.

Reviewer: It might also be helpful to mention here the plan Luke laid out for `wasm32-wasip*-module`. Seems like that would be something realistic that p1-style runtimes could adopt for running newer binaries on older  runtimes?

Reviewer: Or see what I suggested below...

> There has been considerable effort expended to ensure that the WASI 0.1 tool chain remains useable while the WASI 0.2+ transition continues, an example of this is the [WASI Preview Transition](https://github.com/abrown/wasi-libc/blob/preview2-plan/docs/preview2-transition.md). There is as yet no way to run WASI 0.2+ binaries on a WASI 0.1 runtime, although there have been some investigations, like [this one from Marcin](https://github.com/loganek/wasi-snapshot-preview2-to-preview1-adapter), non have been successfully reached a conclusion and been used in production. Additionally, the long term continued investment in WASI 0.1 is not guaranteed. The organizations supporting WASI 0.1 are the same organizations driving WASI 0.2+, as WASI moves towards a 1.0 release these organizations will not have the need to continue to offer support for WASI 0.1.

Reviewer: ```suggestion
There has been considerable effort expended to ensure that the WASI 0.1 tool chain remains useable while the WASI 0.2+ transition continues, an example of this is the [WASI Preview Transition](https://github.com/abrown/wasi-libc/blob/preview2-plan/docs/preview2-transition.md). There is as yet no way to run WASI 0.2+ binaries on a WASI 0.1 runtime and, although there have been some investigations, like [this one from Marcin](https://github.com/loganek/wasi-snapshot-preview2-to-preview1-adapter), none have been successfully reached a conclusion and been used in production. Additionally, the long term continued investment in WASI 0.1 is not guaranteed. The organizations supporting WASI 0.1 are the same organizations driving WASI 0.2+, as WASI moves towards a 1.0 release these organizations will not have the need to continue to offer support for WASI 0.1.
```

Author: Good Catch! :)

> The lack of a commitment to provide long term support for WASI 0.1 and the inability of WASI 0,1 runtimes to execute the newer binary formats produced by the WASI 0.2+ toolchains will result in leaving early adopters without a viable migration path. If left unaddressed this risks fragmenting the ecosystem and undermining trust in WASI’s long-term viability. 

Reviewer: ```suggestion
The lack of a commitment to provide long term support for WASI 0.1 and the inability of WASI 0.1 runtimes to execute the newer binary formats produced by the WASI 0.2+ toolchains will result in leaving early adopters without a viable migration path. If left unaddressed this risks fragmenting the ecosystem and undermining trust in WASI’s long-term viability. 
```

Author: Good catch! - Thank you! :)

> The lack of a commitment to provide long term support for WASI 0.1 and the inability of WASI 0,1 runtimes to execute the newer binary formats produced by the WASI 0.2+ toolchains will result in leaving early adopters without a viable migration path. If left unaddressed this risks fragmenting the ecosystem and undermining trust in WASI’s long-term viability. 

Reviewer: ```suggestion
The inability of WASI 0.1 runtimes to execute the newer binary formats produced by the WASI 0.2+ toolchains could result in leaving early adopters without a viable migration path. Without adopting a proposal like https://github.com/WebAssembly/component-model/pull/378 (or similar), this toolchain gap risks fragmenting the ecosystem and undermining trust in WASI’s long-term viability. 
```

@woodsmc, what do you think of this? I'm trying to highlight that there are proposals out there that would resolve this issue &mdash; it might motivate someone to pick up a shovel to work on solving this.

Author: Yes, it's a good idea (and sorry for the delay on this). In my original I was trying to highlight the urgency to step up. Let's me see if I can combine those two, like an enumerated (none-exhaustive) list of initial efforts.
